### PR TITLE
channeldb+rpc: optimize graph related RPC calls 

### PR DESCRIPTION
**Pull request by Roasbeef**
This PR moves to optimize our two heaviest RPC calls (`DescribeGraph` and `GetNetworkInfo`) by cache either the raw data or using the existing channel graph cache. A new command line flag has been added to enable/disable the cache for `DescribeGraph`. This PR is meant to significantly reduce the memory overhead of running `lndmon` as it tends to hit these RPC calls pretty frequently.

### CODE REVIEW

File: rpcserver.go [hunk 1]
```diff
@@ -5369,6 +5390,21 @@ func (r *rpcServer) DescribeGraph(ctx context.Context,
...
 
+	// Check to see if the cache is already populated, if so then we can
+	// just return it directly.
+	//
+	// TODO(roasbeef): move this to an interceptor level feature?
+	graphCacheActive := r.cfg.Caches.RPCGraphCacheDuration != 0
+	if graphCacheActive {
+		r.graphCache.Lock()
+		if r.describeGraphResp != nil {
+			r.graphCache.Unlock()
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply rpcserver.go:5401]: Moved to add a defer directly after the initial call to lock.

File: rpcserver.go [hunk 2]
```diff
@@ -5751,9 +5748,9 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
 		}
 
 		return nil
-	}); err != nil {
-		return nil, err
-	}
+
+		return nil
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply rpcserver.go:5795]: Fixed!

File: lncfg/graph_caches_rpctest.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+package lncfg
+
+import "time"
+
+var (
+	// DefaultRPCGraphCacheDuration is the default interval that the RPC
+	// response to DescribeGraph should be cached for. When RPC tests are
+	// on, we disable it.
+	DefaultRPCGraphCacheDuration time.Duration = 0
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lncfg/graph_caches_rpctest.go:12]: Nice idea! Adopted

**Comment by Roasbeef** [comment]
Aight just pushed up a new version addressing comments, thanks for the review!

File: channeldb/graph.go [hunk 4]
```diff
@@ -458,6 +458,72 @@ func (c *ChannelGraph) FetchNodeFeatures(
...
+	}
+
+	// Otherwise call back to a version that uses the database directly.
+	// We'll iterate over each node, then the set of channels for each
+	// node, and construct a similar callback functiopn signature as the
+	// main funcotin expects.
+	return c.ForEachNode(func(tx kvdb.RTx, node *LightningNode) error {
+		channels := make(map[uint64]*DirectedChannel)
+
+		err := node.ForEachChannel(tx, func(tx kvdb.RTx,
```
Comment (guggero) [inline channeldb/graph.go:480]: Should we just call into `ForEachNodeChannel` here that already does all this?

File: rpcserver.go [hunk 5]
```diff
@@ -813,6 +817,23 @@ func (r *rpcServer) addDeps(s *server, macService *macaroons.Service,
 	r.chanPredicate = chanPredicate
 	r.macService = macService
 	r.selfNode = selfNode.PubKeyBytes
+
+	graphCacheDuraiton := r.cfg.Caches.RPCGraphCacheDuration
```
Comment (guggero) [inline rpcserver.go:821]: nit: s/graphCacheDuraiton/graphCacheDuration/

**Review by guggero — APPROVED** [review]
LGTM, just one comment about removing some code duplication :tada:

File: channeldb/graph.go (hunk 4 above)
Comment (Roasbeef) [reply channeldb/graph.go:480]: Hmmm, It would need to be slightly different in any case, since I need to map the result to a slightly different call back

**Pull request merged**